app/controllers/concerns/saml_idp_auth_concern.rb

```diff
@@ -3,33 +3,28 @@ module SamlIdpAuthConcern
 
   included do
     before_action :validate_saml_request, only: :auth
-    before_action :validate_service_provider, only: :auth
-    before_action :verify_authn_context, only: :auth
-    before_action :store_saml_request_in_session, only: :auth
+    before_action :validate_service_provider_and_authn_context, only: :auth
+    before_action :store_saml_request_attributes_in_session, only: :auth
     before_action :confirm_two_factor_authenticated, only: :auth
   end
 
   private
 
-  def verify_authn_context
-    return if Saml::Idp::Constants::VALID_AUTHN_CONTEXTS.include?(requested_authn_context)
-
-    process_invalid_authn_context
-  end
-
-  def validate_service_provider
-    add_sp_metadata_to_session and return if current_service_provider.valid?
-
-    analytics.track_event(
-      Analytics::SAML_INVALID_SERVICE_PROVIDER,
-      service_provider: current_service_provider.issuer
+  def validate_service_provider_and_authn_context
+    @result = SamlRequestValidator.new.call(
+      service_provider: current_service_provider,
+      authn_context: requested_authn_context
     )
 
+    return unless @result[:errors].present?
```

Comment:
The errors are captured in the `@result` variable since it's the `SamlRequestValidator` class that processes the validations and errors. If you look at the `saml_idp_controller_spec` changes, you'll see that errors are captured.

This is the same pattern we use in other controllers, where we use a Form Object to process the form and store the result. We then call analytics once and pass in the result hash, then the controller processes the success vs failure scenarios.